Update macOS scripts and support multiple types of GPA (new)

kdk1616 opened the pull request:
This PR contains two features.
* Closes #83. Fix macOS install and start scripts, including removing Redis from the start script.
* Closes #71. Make function for toggling between weighted, unweighted, and % GPA.

kdk1616 committed:
Fixed mac install and start scripts (removed redis from start script). Made function for toggling between weighted, unweighted, and % GPA (only works for current quarter right now... Also, There is a bu where if you select another quarter you cannot return to the initial quarter and have it work properly...).

psvenk committed:
Revert changes to README

Apparently kdk1616's editor completely changed the formatting of
the README. As there were no substantial changes, I have reverted
all changes that were made to the README.

psvenk committed:
Merge branch 'master' into master

kdk1616 committed:
Fixed GPA Toggle not working for other quarters... had to hardcode variable for current term

psvenk committed:
Use function for getting GPA out of 4.0

There is a function in extraFunctions.js to convert a percentage to
a GPA out of 4.0; the inline code duplicating that behavior has been
replaced by a function call.

psvenk committed:
Use tableData.terms.current for current term

This is a minor change, as the currently selected term is the same as
the current term when Aspine is first loading, but it aids code
readability for the refactoring to come.

psvenk committed:
Use for..of loop instead of iterator in GPA calculation

Abstracting away the iterator reduces the potential for errors and
simplifies the code.

psvenk committed:
Improve code readability

psvenk committed:
Further improve code readability and formatting

psvenk committed:
Normalize whitespace and comment out `console.log`s

psvenk committed:
Change GPA to an object + reduce dependence on currentQuarterRightNow

The GPA and calcGPA attributes of each term's object under tableData are
now themselves objects with attributes "percent", "outOfFour", and
"outOfFive". This approach is more scalable and also allows all GPAs
to be returned by a single function, computeGPA, which has been modified
to compute all three formats of GPA.

psvenk committed:
Refactor GPAType so that currentQuarterRightNow is not needed

The current quarter need not be hardcoded anymore.
Also, the refactored GPAType uses jQuery and is more readable.

psvenk committed:
Round 4.0 and 5.0 scale GPAs to 2 places after decimal point

psvenk committed:
Fix bug in which GPA switcher does not work

Because the selection of the <select> element is a string and it is
being compared to the number zero, I have changed triple equals
to double equals in order to invoke type coercion.

kdk1616 committed:
Merge remote-tracking branch 'upstream/master'

kdk1616 committed:
Finished GPA types feature, added element to drop down with cumulative GPAs

kdk1616 committed:
Made login button autofocus, made cum GPA limited to two decimals

psvenk commented:
For some reason, the previous pull request with these changes was closed. For reference, below is the list of issues I found in that PR. I will check them off as they are fixed and tested.
- [x] GPA indicator should be merged with GPA type selector, not with term dropdown
- [x] A GPA of 4.0, 3.0, etc. should be displayed with at least one place after the decimal point
- [x] The non-percent GPA formats do not play well with "Calculated GPA"

psvenk committed:
Make computeGPAQuarter function stateless

computeGPAQuarter was using a global variable instead of using its own
argument; this has now been changed.

psvenk committed:
Remove useless variable curQ

psvenk committed:
Normalize indentation and whitespace in GPAType()

psvenk committed:
Add cumGPA to tableData

psvenk committed:
Store cumulative GPA as numbers rather than as strings

psvenk committed:
Make cumGPA() return an object directly

Instead of taking `type` as a parameter and giving a different value
depending on the type, cumGPA() now returns an object containing all GPA
types.

psvenk committed:
Make "GPA Type" button control all terms

The "GPA Type" button now controls the GPA type of all terms in
the dropdown instead of only the selected one, and the term can be
changed while a non-percent GPA type is selected.

psvenk committed:
Remove word "Percent" to be consistent

psvenk committed:
Move "GPA Type" button out of dropdown

This makes the UI work more intuitively and allows the term dropdown
to resize horizontally as needed.

psvenk committed:
Fix bug relating to fetching terms

Fix a bug in which fetching a new term would overwrite the current
term data

Ruborcalor commented on public/home.js:
```diff
@@ -674,34 +672,180 @@ let termsReset = {};
+                );
+			}
+        }
+      }
```
It would be nice to move these functions from home.js to calculate_grade.js

Ruborcalor commented on public/js/buttonFunctions.js:
```diff
@@ -61,46 +61,50 @@ let resetTableData = function() {
+      document.getElementById('gpa_select').options[1].innerHTML = "Current Quarter GPA: "  + GPA.percent + "<br>Calculated GPA: " + calcGPA.percent;
 
     } else {
+	    tableData.currentTermData.calcGPA = computeGPA(tableData.currentTermData.classes,termConverter.indexOf(currentTerm));
```
Why does computeGPA take two arguments here? It's signature seems to account for a single argument.

line 177 calculate_grade.js
```
function computeGPA(classes) {
 let sum = 0; // Sum of classes' grades
 let count = 0; // Number of classes
 let fourSum = 0; // Sum of classes' grades on 4.0 scale
 let fiveSum = 0; // Sum of classes' grades on 5.0 scale
```

psvenk replied:
I don't know; I just removed the second argument in 7aa8356.

Ruborcalor reviewed (commented):
Overall great work! Two minor comments.

psvenk committed:
Make calculated GPA play nicely with unweighted/weighted

Fix some bugs that would be encountered if a user modifies some
assignments, thus generating a calculated GPA, in any one of
the non-percent GPA types.

psvenk committed:
Make sure that calculated GPA is on separate line from actual GPA

psvenk committed:
Remove redundant argument in computeGPA function call

psvenk committed:
Move functions cumGPA and GPAType to extraFunctions.js

psvenk commented:
I have addressed every issue that I have identified, except for redesigning the cumulative GPA indicator so that it updates along with the quarters in showing percent, unweighted, or weighted instead of just showing all three always. It would also be nice to have a calculated cumulative GPA which uses calcGPA for each quarter (falling back to GPA if needed).

psvenk committed:
Make cumulative GPA act like quarters

Instead of having a rich-text presentation of the cumulative GPA showing
all GPA types at once, make it cycle through the GPA types along with
the quarters (current, Q1, Q2, Q3, Q4).

psvenk committed:
Display GPAs to 2 decimal places and store cumGPA rounded

psvenk committed:
Move cumGPA function definition to calculate_grade.js

psvenk committed:
Make cumGPA take overview as an argument

psvenk reviewed (approved):
Looks good to me

kdk1616 commented:
I approve!
